### ISSUE TITLE
Feature request for using hostname to move and connect to individual node in cluster mode

### ISSUE THREAD

**Issue by peijiehao** [Nordix/hiredis-cluster#239]
Context:

Starting from redis 7.x. two new properties cluster-preferred-endpoint-type and cluster-announce-hostname properties have been added. When cluster-preferred-endpoint-type is set to hostname, Moved and Ask responses contain the hostnames.
When cluster-announce-hostname is set, cluster nodes outputs the hostname along with the Ip. https://redis.io/docs/latest/commands/cluster-nodes/

cluster nodes output:

07c37dfeb235213a872192d90877d0cd55635b91 127.0.0.1:30004@31004,hostname4 slave e7d1eecce10fd6bb5eb35b9f99a514335d9ba9ca 0 1426238317239 4 connected

Problem:

When the nodes in the cluster have announced hostnames. hiredis-cluster should parse the hostnames instead of IPs from cluster nodes output and connect to the nodes via hostnames.
By default, hiredis-cluster runs cluster nodes and uses the IPs to make connections to the individual nodes. This integration will not work when the nodes are behind a proxy/NAT gateway, thus, we are expecting that when cluster-preferred-endpoint-type is set to hostname, on moving and connecting to a different node hiredis-cluster will use hostname instead of IP.

Solution:

Suggesting hiredis-cluster to make connections via hostnames if available and default to IPs if not present

**Comment by zuiderkwast** [comment]
Good point. CLUSTER SLOTS shows hostnames too when set to preferred, but there it replaces the IP if I remember correctly. Did you try to configure hiredis-cluster to use CLUSTER SLOTS instead of NODES, to check if it uses the hostnames in this case?

Regarding IP and NAT-ing, it's also possible configure the NAT-ed IP addresses using cluster-announce-ip. They will be returned to clients.

**Comment by tharunsikhinam** [comment]
> Regarding IP and NAT-ing, it's also possible configure the NAT-ed IP addresses using cluster-announce-ip. They will be returned to clients.

then the assumption is the nodes can use the NAT-ed IPs for intra-cluster communication. 
but that might not always be true. 

https://github.com/redis/redis/issues/7460 There is a feature request for what you are suggesting. but until then, having the client support hostnames gets around this issue.

**Comment by zuiderkwast** [comment]
> When cluster-preferred-endpoint-type is set to hostname, Moved and Ask responses contain the hostnames.

Yes, but when cluster-preferred-endpoint-type is set to 'ip', then MOVED and ASK return the IP, even if cluster-announce-hostname is set.

> When cluster-announce-hostname is set, cluster nodes outputs the hostname along with the Ip. https://redis.io/docs/latest/commands/cluster-nodes/
>
> cluster nodes output:
>
> 07c37dfeb235213a872192d90877d0cd55635b91 127.0.0.1:30004@31004,hostname4 slave
> e7d1eecce10fd6bb5eb35b9f99a514335d9ba9ca 0 1426238317239 4 connected

...

> Suggesting hiredis-cluster to make connections via hostnames if available and default to IPs if not present

I'm not sure. If we connect to a hostname but MOVED still returns IP addresses, we need to know which connection corresponds to the IP address too, so I think we should parse IP and hostname, so we can handle all possible configurations.

@bjosv Do you know if we can handle hostnames in CLUSTER SLOTS?

**Comment by tharunsikhinam** [comment]
> I'm not sure. If we connect to a hostname but MOVED still returns IP addresses, we need to know which connection corresponds to the IP address too, so I think we should parse IP and hostname, so we can handle all possible configurations.

you have a point.  you need both IP and hostname to handle all configurations. 

and thanks for sharing https://github.com/valkey-io/valkey/pull/736 , this is exactly what I was looking for !

**Comment by bjosv** [comment]
Enabling the client to send CLUSTER SLOTS should help you since we support hostnames in CLUSTER SLOTS.
We have this [testcase](https://github.com/Nordix/hiredis-cluster/blob/master/tests/scripts/redirect-with-hostname-test.sh) testing it.

You can enable CLUSTER SLOTS with `redisClusterSetOptionRouteUseSlots(cc)`.
